kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/collections.kt

```diff
@@ -40,17 +42,24 @@ fun <T> Arb.Companion.of(vararg collection: T): Arb<T> = element(*collection)
 fun <A> Arb.Companion.set(gen: Gen<A>, range: IntRange = 0..100): Arb<Set<A>> {
    check(!range.isEmpty())
    check(range.first >= 0)
-   val maxIterations = 1000
+   // we may generate duplicates, but we don't know if the underlying gen has sufficient cardinality
+   // to satisfy our range, so we can try for a while, but must not try for ever
+   // the slippage factor controls how many times we will accept a non unique element before giving up,
+   // which is the number of elements in the target set * slippage
+   val slippage = 10
    return arbitrary {
       val genIter = gen.generate(it).iterator()
       val targetSize = it.random.nextInt(range)
+      val maxMisses = targetSize * slippage
       val set = mutableSetOf<A>()
       var iterations = 0
-      while (iterations++ < maxIterations && set.size < targetSize && genIter.hasNext()) {
+      while (iterations < maxMisses && set.size < targetSize && genIter.hasNext()) {
+         val size = set.size
          set.add(genIter.next().value)
+         if (set.size == size) iterations++
```

Comment:
i think we need to clear the counter when there's an increment in size. one way is to leverage the return value of set.add

```kotlin
val isElementAdded = set.add(genIter.next().value)
iterations = if (isElementAdded) 0 else iterations + 1
```

Reply:
Thanks, good call.

Reply:
oh are we not resetting the counter?